src/UCTSearch.cpp

```diff
@@ -684,12 +687,30 @@ void UCTSearch::ponder() {
         tg.add_task(UCTWorker(m_rootstate, this, m_root.get()));
     }
     auto keeprunning = true;
+    int last_update = 0;
     do {
         auto currstate = std::make_unique<GameState>(m_rootstate);
         auto result = play_simulation(*currstate, m_root.get());
         if (result.valid()) {
             increment_playouts();
         }
+        Time elapsed;
+        int elapsed_centis = Time::timediff_centis(start, elapsed);
+
+        // output some stats every few seconds
+        // check if we should still search
+        if (elapsed_centis - last_update > (cfg_rt_log ? 10 : 250)) {
+            last_update = elapsed_centis;
+
+            if (cfg_rt_log) {
+                myprintf("~begin\n");
+                dump_stats(m_rootstate, *m_root);
```

Comment:
This isn't threadsafe.

Reply:
You need to check if taking the root node lock early on in dump_stats still allows the engine to work. You'll get a performance dip every time you dump, but that's unavoidable.

Reply:
If that doesn't work, you rewrite the "root moves" dumping part of dump_stats into a separate function and call that instead. You can't use the sort_root_moves code there though, you'll have to write some code that gets the top x moves and remember them.